File: crates/nux-capi/smoke/capi_metal_smoke.swift
import Foundation
import CoreGraphics
import ImageIO
import Metal
import QuartzCore
import NuxieRuntimeC

final class CompletionBox {
    let semaphore = DispatchSemaphore(value: 0)
}

final class DecodeBox {
    var pixels: UnsafeMutableRawPointer?
    var pixelCount = 0
    var calls = 0
    var retains = 0
    var releases = 0
    var nestedABI = UInt32(NUX_CAPI_ABI_VERSION)

    deinit { free(pixels) }
}

func retainPixels(_ context: UnsafeMutableRawPointer?) {
    guard let context else { return }
    let box = Unmanaged<DecodeBox>.fromOpaque(context).takeUnretainedValue()
    box.retains += 1
    box.nestedABI = nux_capi_abi_version()
}

func releasePixels(_ context: UnsafeMutableRawPointer?) {
    guard let context else { return }
    Unmanaged<DecodeBox>.fromOpaque(context).takeUnretainedValue().releases += 1
}

func decodeImage(
    _ context: UnsafeMutableRawPointer?,
    _ request: UnsafePointer<NuxImageDecodeRequest>?,
    _ outImage: UnsafeMutablePointer<NuxDecodedImage>?
) -> UInt32 {
    guard let context, let request, let outImage,
          let encodedPointer = request.pointee.encoded.data else {
        return UInt32(NUX_ASSET_CALLBACK_STATUS_FAILED)
    }
    let box = Unmanaged<DecodeBox>.fromOpaque(context).takeUnretainedValue()
    let encoded = Data(bytes: encodedPointer, count: request.pointee.encoded.len)
    guard let source = CGImageSourceCreateWithData(encoded as CFData, nil),
          let image = CGImageSourceCreateImageAtIndex(source, 0, nil) else {
        return UInt32(NUX_ASSET_CALLBACK_STATUS_FAILED)
    }
    let width = image.width
    let height = image.height
    let rowBytes = width * 4
    let count = rowBytes * height
    guard width > 0, height > 0,
          width <= Int(request.pointee.maximum_dimension),
          height <= Int(request.pointee.maximum_dimension),
          count <= request.pointee.maximum_decoded_bytes else {
        return UInt32(NUX_ASSET_CALLBACK_STATUS_FAILED)
    }
    free(box.pixels)
    box.pixels = calloc(1, count)
    box.pixelCount = count
    guard let pixels = box.pixels,
          let colors = CGColorSpace(name: CGColorSpace.sRGB),
          let bitmap = CGContext(
            data: pixels, width: width, height: height, bitsPerComponent: 8,
            bytesPerRow: rowBytes, space: colors,
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue |
                CGBitmapInfo.byteOrder32Big.rawValue
          ) else {
        return UInt32(NUX_ASSET_CALLBACK_STATUS_FAILED)
    }
    bitmap.draw(image, in: CGRect(x: 0, y: 0, width: width, height: height))
    box.calls += 1
    var retained = NuxRetainedBytes()
    retained.struct_size = UInt32(MemoryLayout<NuxRetainedBytes>.size)
    retained.data = UnsafePointer(pixels.assumingMemoryBound(to: UInt8.self))
    retained.len = count
    retained.owner = context
    retained.retain = retainPixels
    retained.release = releasePixels
    var decoded = NuxDecodedImage()
    decoded.struct_size = UInt32(MemoryLayout<NuxDecodedImage>.size)
    decoded.width = UInt32(width)
    decoded.height = UInt32(height)
    decoded.row_bytes = UInt32(rowBytes)
    decoded.pixel_format = UInt32(NUX_PIXEL_FORMAT_RGBA8_PREMULTIPLIED_SRGB)
    decoded.pixels = retained
    outImage.pointee = decoded
    return UInt32(NUX_ASSET_CALLBACK_STATUS_OK)
}

func check(_ condition: @autoclosure () -> Bool, _ message: String) {
    guard condition() else {
        FileHandle.standardError.write(Data("swift-metal-smoke FAILED: \(message)\n".utf8))
        exit(1)
    }
}

func freeResult(_ result: OpaquePointer?, _ expected: UInt32) {
    check(result != nil, "owned result")
    var status = NUX_STATUS_RUNTIME_ERROR.rawValue
    check(nux_capi_result_status(result, &status) == NUX_STATUS_OK.rawValue, "result status")
    check(status == expected, "expected diagnostic status")
    check(nux_capi_result_free(result) == NUX_STATUS_OK.rawValue, "free result")
}

func stringView(_ pointer: UnsafePointer<CChar>, count: Int) -> NuxStringView {
    var view = NuxStringView()
    view.data = pointer
    view.len = count
    return view
}

func copiedString(_ view: NuxStringView) -> String {
    guard let data = view.data else { return "" }
    return String(
        decoding: UnsafeBufferPointer(
            start: UnsafeRawPointer(data).assumingMemoryBound(to: UInt8.self),
            count: view.len
        ),
        as: UTF8.self
    )
}

func importConfigured(
    bytes: Data,
    hooks: inout NuxAppleAssetHooks,
    composed: Bool,
    file: inout OpaquePointer?,
    result: inout OpaquePointer?
) -> UInt32 {
    func call(
        _ config: inout NuxFileImportConfig
    ) -> UInt32 {
        bytes.withUnsafeBytes { rawBytes in
            nux_file_import_configured(
                rawBytes.bindMemory(to: UInt8.self).baseAddress,
                rawBytes.count,
                &config,
                &file,
                &result
            )
        }
    }

    guard composed else {
        return withUnsafePointer(to: &hooks) { hooksPointer in
            var config = NuxFileImportConfig()
            config.struct_size = UInt32(MemoryLayout<NuxFileImportConfig>.size)
            config.apple_assets = hooksPointer
            return call(&config)
        }
    }
    return "bridge".withCString { moduleName in
        "GenericHostChanges".withCString { scriptName in
            "lua".withCString { luaExtension in
                "pixel.png".withCString { imageName in
                    "png".withCString { pngExtension in
                        var host = NuxHostCommandImportConfig()
                        host.struct_size = UInt32(
                            MemoryLayout<NuxHostCommandImportConfig>.size
                        )
                        host.module_name = stringView(moduleName, count: 6)
                        host.max_script_memory_bytes = 64 * 1024 * 1024
                        host.max_script_interrupts_per_callback = 50_000
                        host.max_commands_per_step = 256
                        host.max_value_depth = 32
                        host.max_value_nodes = 4_096
                        host.max_identifier_bytes = 4_096
                        host.max_string_bytes = 1024 * 1024
                        host.max_value_bytes = 4 * 1024 * 1024
                        host.max_command_bytes_per_step = 4 * 1024 * 1024

                        var script = NuxExpectedFileAssetDescriptor()
                        script.struct_size = UInt32(
                            MemoryLayout<NuxExpectedFileAssetDescriptor>.size
                        )
                        script.ordinal = 0
                        script.kind = NUX_FILE_ASSET_KIND_SCRIPT.rawValue
                        script.has_authored_id = 1
                        script.authored_id = 0
                        script.name = stringView(scriptName, count: 18)
                        script.file_extension = stringView(luaExtension, count: 3)
                        script.is_embedded = 1
                        script.has_contents_record = 1

                        var image = NuxExpectedFileAssetDescriptor()
                        image.struct_size = UInt32(
                            MemoryLayout<NuxExpectedFileAssetDescriptor>.size
                        )
                        image.ordinal = 1
                        image.kind = NUX_FILE_ASSET_KIND_IMAGE.rawValue
                        image.has_authored_id = 1
                        image.authored_id = 7
                        image.name = stringView(imageName, count: 9)
                        image.file_extension = stringView(pngExtension, count: 3)
                        image.is_embedded = 1
                        image.has_contents_record = 1
                        image.required_provider_flags = UInt32(
                            NUX_FILE_ASSET_PROVIDER_IMAGE_DECODE
                        )
                        let expected = [script, image]
                        return expected.withUnsafeBufferPointer { expectedBuffer in
                            withUnsafePointer(to: &host) { hostPointer in
                                withUnsafePointer(to: &hooks) { hooksPointer in
                                    var config = NuxFileImportConfig()
                                    config.struct_size = UInt32(
                                        MemoryLayout<NuxFileImportConfig>.size
                                    )
                                    config.host_commands = hostPointer
                                    config.apple_assets = hooksPointer
                                    config.expected_assets = expectedBuffer.baseAddress
                                    config.expected_asset_count = expectedBuffer.count
                                    return call(&config)
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}

guard CommandLine.arguments.count == 2 ||
        (CommandLine.arguments.count == 3 && CommandLine.arguments[2] == "--composed") else {
    FileHandle.standardError.write(Data("usage: capi_metal_smoke <file.riv> [--composed]\n".utf8))
    exit(2)
}

let composed = CommandLine.arguments.count == 3

let bytes = try Data(contentsOf: URL(fileURLWithPath: CommandLine.arguments[1]))
var file: OpaquePointer?
let decoder = DecodeBox()
var hooks = NuxAppleAssetHooks()
hooks.struct_size = UInt32(MemoryLayout<NuxAppleAssetHooks>.size)
hooks.context = Unmanaged.passUnretained(decoder).toOpaque()
hooks.decode_image = decodeImage
hooks.maximum_external_asset_bytes = 64 * 1024 * 1024
hooks.maximum_total_external_asset_bytes = 256 * 1024 * 1024
hooks.maximum_image_dimension = 8192
hooks.maximum_decoded_image_bytes = 256 * 1024 * 1024
hooks.maximum_total_decoded_image_bytes = 512 * 1024 * 1024
var result: OpaquePointer?
check(
    importConfigured(
        bytes: bytes,
        hooks: &hooks,
        composed: composed,
        file: &file,
        result: &result
    ) == NUX_STATUS_OK.rawValue,
    "import"
)
freeResult(result, NUX_STATUS_OK.rawValue)
check(decoder.calls == 1, "one image decode")
check(decoder.retains == 1 && decoder.releases == 1, "balanced decoded pixels")
check(decoder.nestedABI == 0, "callback reentry rejected")
var assetCount = 0
check(nux_file_asset_count(file, &assetCount) == NUX_STATUS_OK.rawValue, "asset count")
check(assetCount == (composed ? 2 : 1), "file asset catalog")
var asset = NuxFileAssetDescriptorView()
asset.struct_size = UInt32(MemoryLayout<NuxFileAssetDescriptorView>.size)
check(nux_file_asset_descriptor(file, 0, &asset) == NUX_STATUS_OK.rawValue, "asset descriptor")
check(
    asset.kind == (composed ? NUX_FILE_ASSET_KIND_SCRIPT.rawValue :
                              NUX_FILE_ASSET_KIND_IMAGE.rawValue),
    "first catalog kind"
)
if composed {
    check(asset.required_provider_flags == 0, "script provider requirements")
    asset.struct_size = UInt32(MemoryLayout<NuxFileAssetDescriptorView>.size)
    check(nux_file_asset_descriptor(file, 1, &asset) == NUX_STATUS_OK.rawValue, "image descriptor")
    check(asset.kind == NUX_FILE_ASSET_KIND_IMAGE.rawValue, "image catalog kind")
}
check(
    asset.required_provider_flags == UInt32(NUX_FILE_ASSET_PROVIDER_IMAGE_DECODE),
    "image decode provider requirement"
)
var artboard: OpaquePointer?
check(nux_artboard_instance_new(file, 0, &artboard) == NUX_STATUS_OK.rawValue, "artboard")
var viewModel: OpaquePointer?
if composed {
    check(
        nux_view_model_instance_new_authored(file, 0, 0, &viewModel) == NUX_STATUS_OK.rawValue,
        "authored view model"
    )
    check(
        nux_artboard_instance_bind_view_model(artboard, viewModel) == NUX_STATUS_OK.rawValue,
        "bind view model"
    )
}
var player: OpaquePointer?
if composed {
    check(
        "HostCommands".withCString { name in
            let view = stringView(name, count: 12)
            return nux_player_new_state_machine_named(artboard, view, &player)
        } == NUX_STATUS_OK.rawValue,
        "state-machine player"
    )
} else {
    check(nux_player_new_static(artboard, &player) == NUX_STATUS_OK.rawValue, "player")
}
check(nux_file_free(file) == NUX_STATUS_OK.rawValue, "release file before player")
check(nux_artboard_instance_free(artboard) == NUX_STATUS_OK.rawValue, "release artboard before player")

var renderer: OpaquePointer?
result = nil
check(nux_renderer_new_metal(4, 3, &renderer, &result) == NUX_STATUS_OK.rawValue, "renderer")
freeResult(result, NUX_STATUS_OK.rawValue)

var rawDevice: UnsafeMutableRawPointer?
result = nil
check(nux_renderer_copy_metal_device(renderer, &rawDevice, &result) == NUX_STATUS_OK.rawValue, "device")
freeResult(result, NUX_STATUS_OK.rawValue)
check(rawDevice != nil, "+1 MTLDevice")
let deviceObject = Unmanaged<AnyObject>.fromOpaque(rawDevice!).takeRetainedValue()
guard let device = deviceObject as? MTLDevice else {
    fatalError("copied pointer is not an MTLDevice")
}

let layer = CAMetalLayer()
layer.device = device
layer.pixelFormat = .bgra8Unorm
layer.framebufferOnly = true
layer.drawableSize = CGSize(width: 4, height: 3)
layer.maximumDrawableCount = 2
layer.allowsNextDrawableTimeout = true
guard let drawable = layer.nextDrawable() else {
    fatalError("configured layer did not vend a drawable")
}

var operation = NuxMetalRenderOperation()
operation.struct_size = UInt32(MemoryLayout<NuxMetalRenderOperation>.size)
operation.drawable_state = UInt32(NUX_METAL_DRAWABLE_STATE_AVAILABLE)
operation.drawable = Unmanaged.passUnretained(drawable).toOpaque()
operation.clear_color = 0xff11_2233
let completion = CompletionBox()
operation.completion_context = Unmanaged.passRetained(completion).toOpaque()
operation.completion_callback = { context in
    guard let context else { return }
    let completion = Unmanaged<CompletionBox>.fromOpaque(context).takeRetainedValue()
    completion.semaphore.signal()
}
var outcome = NuxRendererOutcome()
outcome.struct_size = UInt32(MemoryLayout<NuxRendererOutcome>.size)
result = OpaquePointer(bitPattern: 1)
check(
    nux_renderer_render_player(renderer, player, &operation, &outcome, &result) == NUX_STATUS_OK.rawValue,
    "render"
)
check(result == nil, "success render performs no diagnostic allocation")
check(outcome.disposition == UInt32(NUX_RENDERER_DISPOSITION_PRESENTED), "presented")
check(
    completion.semaphore.wait(timeout: .now() + 5) == .success,
    "deferred Metal completion"
)

if composed {
    var mutationResult: OpaquePointer?
    let mutationStatus = "amount".withCString { path in
        var mutation = NuxViewModelMutation()
        mutation.kind = NUX_VIEW_MODEL_MUTATION_KIND_SET_NUMBER.rawValue
        mutation.instance = viewModel
        mutation.path = stringView(path, count: 6)
        mutation.number_value = 5
        return withUnsafePointer(to: &mutation) { mutationPointer in
            var batch = NuxViewModelMutationBatch()
            batch.struct_size = UInt32(MemoryLayout<NuxViewModelMutationBatch>.size)
            batch.mutations = mutationPointer
            batch.mutation_count = 1
            batch.correlation_id = 43
            return nux_view_model_mutate(&batch, &mutationResult)
        }
    }
    check(mutationStatus == NUX_STATUS_OK.rawValue, "caller mutation")
    var mutationInfo = NuxViewModelMutationResultInfo()
    mutationInfo.struct_size = UInt32(MemoryLayout<NuxViewModelMutationResultInfo>.size)
    check(
        nux_view_model_mutation_result_info(mutationResult, &mutationInfo) == NUX_STATUS_OK.rawValue,
        "mutation info"
    )
    check(mutationInfo.applied_count == 1, "one applied caller mutation")
    check(mutationInfo.correlation_id == 43, "caller correlation")
    check(mutationInfo.change_count == 1, "one caller journal entry")
    var callerChange = NuxViewModelChangeView()
    callerChange.struct_size = UInt32(MemoryLayout<NuxViewModelChangeView>.size)
    check(
        nux_view_model_mutation_result_change(mutationResult, 0, &callerChange) == NUX_STATUS_OK.rawValue,
        "caller journal"
    )
    check(callerChange.origin == NUX_VIEW_MODEL_CHANGE_ORIGIN_CALLER.rawValue, "caller origin")
    check(callerChange.correlation_id == 43, "caller journal correlation")
    check(callerChange.kind == NUX_VIEW_MODEL_VALUE_KIND_NUMBER.rawValue, "caller number kind")
    check(callerChange.number_value == 5, "caller number value")
    check(
        nux_view_model_mutation_result_free(mutationResult) == NUX_STATUS_OK.rawValue,
        "free mutation journal"
    )

    var initial = NuxPlayerStep()
    initial.struct_size = UInt32(MemoryLayout<NuxPlayerStep>.size)
    var stepResult: OpaquePointer?
    check(nux_player_step(player, &initial, &stepResult) == NUX_STATUS_OK.rawValue, "initialize player")
    check(nux_player_step_result_free(stepResult) == NUX_STATUS_OK.rawValue, "free initial step")

    var down = NuxPlayerPointerEvent()
    down.kind = NUX_PLAYER_POINTER_KIND_DOWN.rawValue
    down.x = 50
    down.y = 50
    var up = NuxPlayerPointerEvent()
    up.kind = NUX_PLAYER_POINTER_KIND_UP.rawValue
    up.x = 50
    up.y = 50
    let pointers = [down, up]
    stepResult = nil
    let stepStatus = pointers.withUnsafeBufferPointer { pointerBuffer in
        var step = NuxPlayerStep()
        step.struct_size = UInt32(MemoryLayout<NuxPlayerStep>.size)
        step.pointers = pointerBuffer.baseAddress
        step.pointer_count = pointerBuffer.count
        step.elapsed_seconds = 0.016
        step.correlation_id = 44
        return nux_player_step(player, &step, &stepResult)
    }
    check(stepStatus == NUX_STATUS_OK.rawValue, "scripted player step")
    var stepInfo = NuxPlayerStepInfo()
    stepInfo.struct_size = UInt32(MemoryLayout<NuxPlayerStepInfo>.size)
    check(
        nux_player_step_result_info(stepResult, &stepInfo) == NUX_STATUS_OK.rawValue,
        "step info"
    )
    check(stepInfo.host_command_count == 1, "one host command")
    check(stepInfo.view_model_change_count == 2, "two ordered runtime changes")
    var command = NuxHostCommandView()
    command.struct_size = UInt32(MemoryLayout<NuxHostCommandView>.size)
    check(
        nux_player_step_result_host_command(stepResult, 0, &command) == NUX_STATUS_OK.rawValue,
        "host command"
    )
    check(copiedString(command.name) == "performed", "host command name")
    for (index, expected) in [Float(10), Float(20)].enumerated() {
        var change = NuxViewModelChangeView()
        change.struct_size = UInt32(MemoryLayout<NuxViewModelChangeView>.size)
        check(
            nux_player_step_result_view_model_change(stepResult, index, &change) == NUX_STATUS_OK.rawValue,
            "runtime journal \(index)"
        )
        check(change.origin == NUX_VIEW_MODEL_CHANGE_ORIGIN_RUNTIME.rawValue, "runtime origin")
        check(change.correlation_id == 44, "runtime correlation")
        check(change.kind == NUX_VIEW_MODEL_VALUE_KIND_NUMBER.rawValue, "runtime number kind")
        check(change.number_value == expected, "ordered runtime number \(index)")
    }
    check(nux_player_step_result_free(stepResult) == NUX_STATUS_OK.rawValue, "free scripted step")
}

check(nux_renderer_free(renderer) == NUX_STATUS_OK.rawValue, "free renderer")
check(nux_player_free(player) == NUX_STATUS_OK.rawValue, "free player")
check(nux_view_model_instance_free(viewModel) == NUX_STATUS_OK.rawValue, "free view model")
check(decoder.calls == 1, "render does not call host decoder again")
print("swift-metal-smoke ok")

File: crates/nux-capi/smoke/capi_metal_smoke.c
/* Product-neutral Apple renderer C ABI smoke. The C program uses only the C
 * ABI (including the Objective-C C runtime) to configure a CAMetalLayer,
 * render a retained drawable, and exercise the complete handle lifecycle.
 *
 * Usage: capi_metal_smoke <path-to-in_band_asset.riv>
 */

#include "nux_capi_apple.h"

#include <CoreGraphics/CoreGraphics.h>
#include <ImageIO/ImageIO.h>
#include <dispatch/dispatch.h>
#include <objc/message.h>
#include <objc/runtime.h>
#include <stdio.h>
#include <stdatomic.h>
#include <stdlib.h>

/* MTLPixelFormatBGRA8Unorm's stable Objective-C ABI value. Metal's public
 * header is Objective-C-only, while this host intentionally remains C11. */
#define MTL_PIXEL_FORMAT_BGRA8_UNORM 80UL

#define CHECK(condition)                                                      \
    do                                                                        \
    {                                                                         \
        if (!(condition))                                                     \
        {                                                                     \
            fprintf(stderr, "capi-metal-smoke FAILED at %s:%d: %s\n",        \
                    __FILE__, __LINE__, #condition);                          \
            exit(1);                                                          \
        }                                                                     \
    } while (0)

static uint8_t* read_file(const char* path, size_t* out_len)
{
    FILE* file = fopen(path, "rb");
    CHECK(file != NULL);
    CHECK(fseek(file, 0, SEEK_END) == 0);
    long size = ftell(file);
    CHECK(size > 0);
    CHECK(fseek(file, 0, SEEK_SET) == 0);
    uint8_t* bytes = malloc((size_t)size);
    CHECK(bytes != NULL);
    CHECK(fread(bytes, 1, (size_t)size, file) == (size_t)size);
    CHECK(fclose(file) == 0);
    *out_len = (size_t)size;
    return bytes;
}

static void free_result(NuxCapiResult* result, NuxStatus expected)
{
    CHECK(result != NULL);
    NuxStatus status = NUX_STATUS_RUNTIME_ERROR;
    CHECK(nux_capi_result_status(result, &status) == NUX_STATUS_OK);
    CHECK(status == expected);
    CHECK(nux_capi_result_free(result) == NUX_STATUS_OK);
}

typedef void* ObjcObject;

typedef struct DecodeProbe
{
    uint8_t* pixels;
    size_t pixel_len;
    unsigned calls;
    unsigned retains;
    unsigned releases;
    uint32_t nested_abi;
} DecodeProbe;

static void retain_pixels(void* owner)
{
    DecodeProbe* probe = owner;
    probe->retains += 1;
    probe->nested_abi = nux_capi_abi_version();
}

static void release_pixels(void* owner)
{
    DecodeProbe* probe = owner;
    probe->releases += 1;
}

static NuxAssetCallbackStatus decode_image(void* context,
                                           const NuxImageDecodeRequest* request,
                                           NuxDecodedImage* out_image)
{
    DecodeProbe* probe = context;
    CHECK(request != NULL);
    CHECK(request->struct_size >= NUX_IMAGE_DECODE_REQUEST_V3_MIN_SIZE);
    CFDataRef encoded = CFDataCreateWithBytesNoCopy(
        kCFAllocatorDefault, request->encoded.data, (CFIndex)request->encoded.len,
        kCFAllocatorNull);
    CHECK(encoded != NULL);
    CGImageSourceRef source = CGImageSourceCreateWithData(encoded, NULL);
    CHECK(source != NULL);
    CGImageRef image = CGImageSourceCreateImageAtIndex(source, 0, NULL);
    CHECK(image != NULL);
    size_t width = CGImageGetWidth(image);
    size_t height = CGImageGetHeight(image);
    CHECK(width > 0 && height > 0);
    CHECK(width <= request->maximum_dimension &&
          height <= request->maximum_dimension);
    CHECK(width <= UINT32_MAX / 4);
    size_t row_bytes = width * 4;
    CHECK(height <= SIZE_MAX / row_bytes);
    size_t pixel_len = row_bytes * height;
    CHECK(pixel_len <= request->maximum_decoded_bytes);
    free(probe->pixels);
    probe->pixels = calloc(1, pixel_len);
    CHECK(probe->pixels != NULL);
    probe->pixel_len = pixel_len;
    CGColorSpaceRef colors = CGColorSpaceCreateWithName(kCGColorSpaceSRGB);
    CHECK(colors != NULL);
    CGContextRef bitmap = CGBitmapContextCreate(
        probe->pixels, width, height, 8, row_bytes, colors,
        kCGImageAlphaPremultipliedLast | kCGBitmapByteOrder32Big);
    CHECK(bitmap != NULL);
    CGContextDrawImage(bitmap, CGRectMake(0, 0, (CGFloat)width, (CGFloat)height), image);
    CGContextRelease(bitmap);
    CGColorSpaceRelease(colors);
    CGImageRelease(image);
    CFRelease(source);
    CFRelease(encoded);
    probe->calls += 1;
    *out_image = (NuxDecodedImage){
        .struct_size = sizeof(NuxDecodedImage),
        .width = (uint32_t)width,
        .height = (uint32_t)height,
        .row_bytes = (uint32_t)row_bytes,
        .pixel_format = NUX_PIXEL_FORMAT_RGBA8_PREMULTIPLIED_SRGB,
        .pixels = {
            .struct_size = sizeof(NuxRetainedBytes),
            .data = probe->pixels,
            .len = probe->pixel_len,
            .owner = probe,
            .retain = retain_pixels,
            .release = release_pixels,
        },
    };
    return NUX_ASSET_CALLBACK_STATUS_OK;
}

static ObjcObject send_object(ObjcObject receiver, const char* selector)
{
    return ((ObjcObject(*)(ObjcObject, SEL))objc_msgSend)(
        receiver, sel_registerName(selector));
}

static void send_object_argument(ObjcObject receiver,
                                 const char* selector,
                                 ObjcObject argument)
{
    ((void (*)(ObjcObject, SEL, ObjcObject))objc_msgSend)(
        receiver, sel_registerName(selector), argument);
}

static void send_unsigned_argument(ObjcObject receiver,
                                   const char* selector,
                                   unsigned long argument)
{
    ((void (*)(ObjcObject, SEL, unsigned long))objc_msgSend)(
        receiver, sel_registerName(selector), argument);
}

static void send_bool_argument(ObjcObject receiver,
                               const char* selector,
                               signed char argument)
{
    ((void (*)(ObjcObject, SEL, signed char))objc_msgSend)(
        receiver, sel_registerName(selector), argument);
}

static void send_size_argument(ObjcObject receiver,
                               const char* selector,
                               CGSize argument)
{
    ((void (*)(ObjcObject, SEL, CGSize))objc_msgSend)(
        receiver, sel_registerName(selector), argument);
}

static void send_void(ObjcObject receiver, const char* selector)
{
    ((void (*)(ObjcObject, SEL))objc_msgSend)(
        receiver, sel_registerName(selector));
}

typedef struct CompletionProbe
{
    dispatch_semaphore_t semaphore;
    atomic_bool call_is_active;
    atomic_bool called_inline;
    atomic_uint calls;
} CompletionProbe;

static void render_completed(void* context)
{
    CompletionProbe* probe = context;
    if (atomic_load_explicit(&probe->call_is_active, memory_order_acquire))
    {
        atomic_store_explicit(&probe->called_inline, true, memory_order_release);
    }
    atomic_fetch_add_explicit(&probe->calls, 1, memory_order_release);
    dispatch_semaphore_signal(probe->semaphore);
}

int main(int argc, char** argv)
{
    CHECK(argc == 2);
    size_t len = 0;
    uint8_t* bytes = read_file(argv[1], &len);
    DecodeProbe decoder = {0};
    NuxAppleAssetHooks hooks = {
        .struct_size = sizeof(NuxAppleAssetHooks),
        .context = &decoder,
        .decode_image = decode_image,
        .maximum_external_asset_bytes = 64 * 1024 * 1024,
        .maximum_total_external_asset_bytes = 256 * 1024 * 1024,
        .maximum_image_dimension = 8192,
        .maximum_decoded_image_bytes = 256 * 1024 * 1024,
        .maximum_total_decoded_image_bytes = 512 * 1024 * 1024,
    };
    NuxFileImportConfig import_config = {
        .struct_size = sizeof(NuxFileImportConfig),
        .apple_assets = &hooks,
    };
    NuxFile* file = NULL;
    NuxCapiResult* result = NULL;
    CHECK(nux_file_import_configured(
              bytes, len, &import_config, &file, &result) == NUX_STATUS_OK);
    free_result(result, NUX_STATUS_OK);
    CHECK(decoder.calls == 1);
    CHECK(decoder.retains == 1);
    CHECK(decoder.releases == 1);
    CHECK(decoder.nested_abi == 0);
    size_t asset_count = 0;
    CHECK(nux_file_asset_count(file, &asset_count) == NUX_STATUS_OK);
    CHECK(asset_count == 1);
    NuxFileAssetDescriptorView asset = {
        .struct_size = sizeof(NuxFileAssetDescriptorView)};
    CHECK(nux_file_asset_descriptor(file, 0, &asset) == NUX_STATUS_OK);
    CHECK(asset.kind == NUX_FILE_ASSET_KIND_IMAGE);
    CHECK(asset.required_provider_flags == NUX_FILE_ASSET_PROVIDER_IMAGE_DECODE);
    free(bytes);
    NuxArtboardInstance* artboard = NULL;
    CHECK(nux_artboard_instance_new(file, 0, &artboard) == NUX_STATUS_OK);
    NuxPlayer* player = NULL;
    CHECK(nux_player_new_static(artboard, &player) == NUX_STATUS_OK);
    CHECK(nux_file_free(file) == NUX_STATUS_OK);
    CHECK(nux_artboard_instance_free(artboard) == NUX_STATUS_OK);

    NuxRenderer* renderer = NULL;
    result = NULL;
    CHECK(nux_renderer_new_metal(4, 3, &renderer, &result) == NUX_STATUS_OK);
    free_result(result, NUX_STATUS_OK);

    void* device = NULL;
    result = NULL;
    CHECK(nux_renderer_copy_metal_device(renderer, &device, &result) ==
          NUX_STATUS_OK);
    free_result(result, NUX_STATUS_OK);
    CHECK(device != NULL);

    ObjcObject layer_class = (ObjcObject)objc_getClass("CAMetalLayer");
    CHECK(layer_class != NULL);
    ObjcObject layer = send_object(layer_class, "layer");
    CHECK(layer != NULL);
    send_object_argument(layer, "setDevice:", device);
    send_unsigned_argument(
        layer, "setPixelFormat:", MTL_PIXEL_FORMAT_BGRA8_UNORM);
    send_bool_argument(layer, "setFramebufferOnly:", 1);
    send_size_argument(layer, "setDrawableSize:", CGSizeMake(4.0, 3.0));
    send_unsigned_argument(layer, "setMaximumDrawableCount:", 2);
    send_bool_argument(layer, "setAllowsNextDrawableTimeout:", 1);
    send_void(device, "release");

    ObjcObject drawable = send_object(layer, "nextDrawable");
    CHECK(drawable != NULL);
    drawable = send_object(drawable, "retain");
    CHECK(drawable != NULL);
    CompletionProbe completion = {
        .semaphore = dispatch_semaphore_create(0),
        .call_is_active = ATOMIC_VAR_INIT(false),
        .called_inline = ATOMIC_VAR_INIT(false),
        .calls = ATOMIC_VAR_INIT(0),
    };
    CHECK(completion.semaphore != NULL);
    NuxMetalRenderOperation operation = {
        .struct_size = sizeof(NuxMetalRenderOperation),
        .drawable_state = NUX_METAL_DRAWABLE_STATE_AVAILABLE,
        .drawable = drawable,
        .clear_color = 0xff112233,
        .completion_context = &completion,
        .completion_callback = render_completed,
    };
    NuxRendererOutcome outcome = {.struct_size = sizeof(NuxRendererOutcome)};
    result = (NuxCapiResult*)1;
    atomic_store_explicit(
        &completion.call_is_active, true, memory_order_release);
    CHECK(nux_renderer_render_player(
              renderer, player, &operation, &outcome, &result) == NUX_STATUS_OK);
    atomic_store_explicit(
        &completion.call_is_active, false, memory_order_release);
    send_void(drawable, "release");
    CHECK(result == NULL);
    CHECK(outcome.disposition == NUX_RENDERER_DISPOSITION_PRESENTED);
    CHECK(dispatch_semaphore_wait(
              completion.semaphore,
              dispatch_time(DISPATCH_TIME_NOW, 5 * NSEC_PER_SEC)) == 0);
    CHECK(atomic_load_explicit(&completion.calls, memory_order_acquire) == 1);
    CHECK(!atomic_load_explicit(
        &completion.called_inline, memory_order_acquire));

    result = NULL;
    CHECK(nux_renderer_resize(renderer, 0, 0, &outcome, &result) == NUX_STATUS_OK);
    free_result(result, NUX_STATUS_OK);
    CHECK(outcome.disposition == NUX_RENDERER_DISPOSITION_SKIPPED_ZERO_SIZE);
    result = NULL;
    CHECK(nux_renderer_detach(renderer, &outcome, &result) == NUX_STATUS_OK);
    free_result(result, NUX_STATUS_OK);
    result = NULL;
    CHECK(nux_renderer_reattach(renderer, 4, 3, &outcome, &result) == NUX_STATUS_OK);
    free_result(result, NUX_STATUS_OK);
    CHECK(outcome.disposition == NUX_RENDERER_DISPOSITION_RECREATED);

    CHECK(nux_renderer_free(renderer) == NUX_STATUS_OK);
    CHECK(nux_player_free(player) == NUX_STATUS_OK);
    CHECK(decoder.calls == 1);
    free(decoder.pixels);
    puts("capi-metal-smoke ok");
    return 0;
}
